artichoke-backend/src/extn/core/time/offset.rs

```diff
@@ -0,0 +1,211 @@
+        let mut in_value = hash.get(0).expect("Only the `in` parameter should be available").1;
+
+        match in_value.ruby_type() {
+            Ruby::String => {
+                let offset_str = unsafe { implicitly_convert_to_string(self, &mut in_value) }?;
+
+                let offset = Offset::try_from(offset_str).map_err(|_| {
+                    let mut message =
+                        b"\"+HH:MM\", \"-HH:MM\", \"UTC\" or \"A\"..\"I\",\"K\"..\"Z\" expected for utc_offset: "
+                            .to_vec();
+                    message.extend_from_slice(offset_str);
+                    ArgumentError::from(message)
+                })?;
+
+                Ok(Some(offset))
+            }
+            Ruby::Float => {
+                // This impl differs from MRI. MRI supports any float value and
+                // will set an offset with subsec fractions however this is not
+                // supported in spinoso_time with the `tzrs` feature.
+                let offset_seconds: f64 = self.try_convert(in_value)?;
+
+                if (MIN_FLOAT_OFFSET..=MAX_FLOAT_OFFSET).contains(&offset_seconds) {
+                    #[allow(clippy::cast_possible_truncation)]
+                    Ok(Some(Offset::try_from(offset_seconds as i32)?))
+                } else {
+                    Err(ArgumentError::with_message("utc_offset out of range").into())
+                }
+            }
+            _ => {
+                let offset_seconds = implicitly_convert_to_int(self, in_value).and_then(|seconds| {
+                    i32::try_from(seconds).map_err(|_| ArgumentError::with_message("utc_offset out of range").into())
+                })?;
+
+                Ok(Some(Offset::try_from(offset_seconds)?))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::extn::core::time::Offset;
+    use crate::test::prelude::*;
+
+    #[test]
+    fn no_options_does_not_raise() {
+        let mut interp = interpreter();
+
+        let options = interp.eval(b"{}").unwrap();
+
+        let offset: Option<Offset> = interp.try_convert_mut(options).unwrap();
+        assert_eq!(offset, None);
+    }
+
+    #[test]
+    fn raises_on_keys_except_in() {
+        let mut interp = interpreter();
+
+        let options = interp.eval(b"{ foo: 'bar' }").unwrap();
+
+        let result: Result<Option<Offset>, Error> = interp.try_convert_mut(options);
+        let error = result.unwrap_err();
+
+        assert_eq!(error.name(), "ArgumentError");
+        assert_eq!(error.message(), b"unknown keyword: foo".as_slice());
+    }
+
+    #[test]
+    fn raises_on_invalid_timezone_string() {
+        let mut interp = interpreter();
+
+        let options = interp.eval(b"{ in: 'J' }").unwrap();
+
+        let result: Result<Option<Offset>, Error> = interp.try_convert_mut(options);
+        let error = result.unwrap_err();
+
+        assert_eq!(error.name(), "ArgumentError");
+        assert_eq!(
+            error.message(),
```

Comment:
turning byte strings into `BStr`s with `.as_bstr()` makes the `assert_eq!` error messages easier to read by printing out a UTF-8-ish form instead of a sequence of `u8` bytes.